File: assignments/ruby/pythagorean-triplet/pythagorean-triplet_test.rb
require 'minitest/autorun'
require_relative 'triplet'

class TripletTest < MiniTest::Unit::TestCase
  def test_sum
    assert_equal 12, Triplet.new(3, 4, 5).sum
  end

  def test_product
    skip
    assert_equal 60, Triplet.new(3, 4, 5).product
  end

  def test_pythagorean
    skip
    assert Triplet.new(3, 4, 5).pythagorean?
  end

  def test_not_pythagorean
    skip
    assert !Triplet.new(5, 6, 7).pythagorean?
  end

  def test_triplets_upto_10
    skip
    triplets = Triplet.where(max_factor: 10)
    products = triplets.map(&:product).sort
    assert_equal [60, 480], products
  end

  def test_triplets_from_11_upto_20
    skip
    triplets = Triplet.where(min_factor: 11, max_factor: 20)
    products = triplets.map(&:product).sort
    assert_equal [3840], products
  end

  def test_triplets_where_sum_x
    skip
    triplets = Triplet.where(sum: 180, max_factor: 100)
    products = triplets.map(&:product).sort
    assert_equal [118080, 168480, 202500], products
  end
end

File: assignments/elixir/pythagorean-triplet/pythagorean_triplet_test.exs
Code.load_file("pythagorean_triplet.exs")
ExUnit.start

defmodule PythagoreanTripletTest do
  use ExUnit.Case, async: true

  test "sum" do
    triplet = [3, 4, 5]
    assert 12 == Triplet.sum(triplet)
  end

  test "product" do
    triplet = [3, 4, 5]
    assert 60 == Triplet.product(triplet)
  end

  test "pythagorean" do
    triplet = [3, 4, 5]
    assert Triplet.pythagorean?(triplet)
  end

  test "not pythagorean" do
    triplet = [5, 6, 7]
    assert !Triplet.pythagorean?(triplet)
  end

  test "triplets up to 10" do
    triplets = Triplet.generate(10)
    assert [60, 480] == Enum.map(triplets, &(Triplet.product(&1)))
  end

  test "triplets from 11 up to 20" do
    triplets = Triplet.generate(11, 20)
    assert [3840] == Enum.map(triplets, &(Triplet.product(&1)))
  end

  test "triplets where sum is 180 and max factor is 100" do
    triplets = Triplet.generate(1, 100, 180)
    assert [118080, 168480, 202500] == Enum.map(triplets, &(Triplet.product(&1)))
  end


end